Development

pruthvirajmv opened the pull request:
Hello,
Here is my initial version of the E-commerce app.
Kindly review and provide your feedback.
For now not created a landing page & product page as they are most related to CSS😅

pruthvirajmv committed:
initial commit

pruthvirajmv committed:
Merge pull request #1 from pruthvirajmv/csb-f6iku

update: initial release

pruthvirajmv committed:
update: ReadMe

pruthvirajmv committed:
update: CSS lib link update

pruthvirajmv committed:
fix: cart items 
fixed UI for the mobile view of the cart items

supminn commented on src/App.js:
```diff
@@ -1,10 +1,77 @@
+
+  const itemsInCart = state.ProductsList.filter((item) => item.quantity > 0);
+
+  function AppNavBar() {
+    return (
+      <>
+        <nav class="nav nav-dark">
+          <button
+            className="bttn-no-style brand-name"
+            onClick={() => setRoute(routes.Products)}
+          >
+            <h3>E-Commerce</h3>
+          </button>
+          <ul class="nav-list">
+            <li key="products" onClick={() => setRoute(routes.Products)}>
+              Products
+            </li>
+            <li key="wishlist" onClick={() => setRoute(routes.Wishilists)}>
+              <div class="badge">
+                <i class="fa fa-heart" aria-hidden="true"></i>
+                <span class="badge badge-icon">{itemsInWishList.length}</span>
+              </div>
+            </li>
+            <li key="Cart" onClick={() => setRoute(routes.CartItems)}>
+              <div class="badge">
+                <i class="fa fa-shopping-cart " aria-hidden="true"></i>
+                <span class="badge badge-icon">{itemsInCart.length}</span>
+              </div>
+            </li>
+          </ul>
+        </nav>
+      </>
+    );
+  }
```
This can be moved into a new file within the components folder.

pruthvirajmv replied:
sure will do

supminn commented on src/api/mock.commerce.server.js:
```diff
@@ -0,0 +1,37 @@
+          adjective: faker.commerce.productAdjective(),
+          image: faker.random.image(),
+          price: faker.commerce.price(),
+          quantity: 0,
```
if "quantity" is being used only for cart items and not products in general, then you can omit it here and add it within the cart array.

pruthvirajmv replied:
"quantity" has been used in all the pages and as of now, I do not have a cart array in my database or in the state. And have realized there should be a separate cart & wishlist array and will modify the DB in next version

pruthvirajmv committed:
Update README.md

supminn commented on src/Components/useRoute.js:
```diff
@@ -0,0 +1,14 @@
+import { useState } from "react";
+
+export const routes = {
+  Products: "ProductsListingPage",
+  Wishilists: "WishlistPage",
+  CartItems: "CartPage",
```
you can omit repetition of the word "page".

pruthvirajmv replied:
this for readability and anyway this routes will be eliminated when I use Router in next version

supminn commented on src/Components/commerce-context/commerce-context.js:
```diff
@@ -0,0 +1,56 @@
+  const [showToast, setShowToast] = useState("Hide");
+
+  useEffect(() => {
+    setIsLoading("fetchStarted");
+    (async () => {
+      try {
+        const response = await axios.get("/api/products");
```
As a better practice, setIsLoading can be added within the 'try' block.

pruthvirajmv replied:
sure

supminn commented on src/Components/commerce-context/commerce-context.js:
```diff
@@ -0,0 +1,56 @@
+    setShowToast("Show");
+    let showToastTimer = setTimeout(() => setShowToast("Hide"), 1100);
+    return () => clearTimeout(showToastTimer);
+  }, [state, state.ToastMsg]);
+
+  return (
+    <CommerceContext.Provider
+      value={{ state, dispatch, reducer, isLoading, showToast, setShowToast }}
+    >
```
when toastMsg is already a part of state, why do we need to create a new state to show/hide toast? we could manage the same using the dispatch right?

pruthvirajmv replied:
nothing specifically. This make sense, will improvise here

pruthvirajmv replied:
handled the toast & toast msg inside the state itself

supminn commented on src/Components/commerce-context/commerce-reducer.js:
```diff
@@ -0,0 +1,98 @@
+import { toastMsgs } from "../Functionalities/toastMsgs";
+
+export default function reducer(state, action) {
+  switch (action.type) {
```
It is recommended to de structure the action object so that type and payload could be individually used without the repetition of "action" in each instance.

pruthvirajmv replied:
true, thanks for pointing it out

supminn commented on src/Components/commerce-context/commerce-reducer.js:
```diff
@@ -0,0 +1,98 @@
+        ToastMsg: !action.payload.isWishListed
+          ? toastMsgs.addToWishList
+          : toastMsgs.removeFromWishlist
+      };
```
This is a nice way to create the logic!

supminn commented on src/Components/commerce-context/commerce-reducer.js:
```diff
@@ -0,0 +1,98 @@
+        ...state,
+        ProductsList: state.ProductsList.map((item) =>
+          item.id === action.payload.id
+            ? { ...item, isWishListed: true, quantity: 0 }
```
Do we need to add quantity: 0 here? this property is present by default within productlist data right?

pruthvirajmv replied:
as I mentioned before, currently in DB there is no separate cart data. Everything is handled by productlist. 
need to modify the data model

supminn commented on src/styles.css:
```diff
@@ -1,4 +1,119 @@
+}
+
+.card {
+  width: 250px;
```
You could use rem instead of px.

pruthvirajmv replied:
yea sure

supminn commented on src/Components/Functionalities/toastMsgs.js:
```diff
@@ -0,0 +1,8 @@
+export const toastMsgs = {
+  addToCart: "Added To Cart",
+  cartUpdate: "Cart Updated",
+  removeFromCart: "Removed from Cart",
+  addToWishList: "Added to Wishlist",
+  removeFromWishlist: "Removed from Wishlist",
+  moveToWishlist: "Moved to Wishlist"
+};
```
This is a good practice! Nevertheless, uppercase is preferred as a standard practice.
Ex: ADD_TO_CART

pruthvirajmv replied:
Those msgs are for the toast to display not for the switch case in the reducer function

supminn commented on src/Components/Functionalities/GoToCartButton.js:
```diff
@@ -0,0 +1,15 @@
+    <>
+      <button
+        onClick={() => setRoute(routes.CartItems)}
+        style={{ marginLeft: "1rem" }}
```
Try to avoid inline styles. Instead, you could use a className and add the styles within App.css

pruthvirajmv replied:
sure will do

supminn commented on src/Components/WishListPage/WishListPage.js:
```diff
@@ -0,0 +1,64 @@
+      <>
+        <div className="products-display">
+          {wishListedItems.map((item) => (
+            <div>
```
I think this div can be omitted.

pruthvirajmv replied:
yes,will look into it

supminn commented on src/Components/WishListPage/WishListPage.js:
```diff
@@ -0,0 +1,64 @@
+    );
+  }
+
+  return (
+    <>
+      {wishListedItems.length === 0 && (
+        <div className="empty-msg">Wishlist is Empty</div>
+      )}
+      <ProductsListWishListPage />
+    </>
+  );
+}
```
I would suggest adding the code from line 58 to 60 on line number 16.

pruthvirajmv replied:
ok sure

agauravdev commented on src/Components/Functionalities/IncrementDecrementButton.js:
```diff
@@ -0,0 +1,31 @@
+export default function IncrementDecrementButton({ clickHandler, value }) {
```
It's so good that you abstracted this file out too.. I really love this. Gonna steal it.. Huhuhahaha.. xD

agauravdev commented on src/Components/ProductListing/ProductListing.js:
```diff
@@ -0,0 +1,98 @@
+    searchProducts: ""
+  });
+
+  function filterData(
```
Can put these functions in their own respectives files in a folder named "helpers" or "utils"? Idk, just a thought.

pruthvirajmv replied:
sure will do

agauravdev reviewed (commented):
Only some minor changes. Nothing major. Amazing code dude, the code was clean and readable. Gonna steal some good practices from here.. huhu :P

supminn reviewed (commented):
Hi @pruthvirajmv, very well organized code, I have added a few minor comments on following better practices.

pruthvirajmv committed:
refactor: per PR
refactored code & styles as PR review

pruthvirajmv committed:
refactor

pruthvirajmv committed:
feat/refactor: productpage, router / components
created product page and implemented router in the app along with the code refactoring

pruthvirajmv committed:
app features status

pruthvirajmv committed:
Update: Readme correction

pruthvirajmv committed:
styles: favicon, product hover
added redirects
